Std/Data/List/Count.lean

```diff
@@ -0,0 +1,260 @@
+  apply s.filter _ |>.length_le
+
+@[simp] theorem countp_filter (l : List α) :
+    countp p (filter q l) = countp (fun a => p a ∧ q a) l := by
+  simp only [countp_eq_length_filter, filter_filter]
+
+@[simp] theorem countp_true {l : List α} : (l.countp fun _ => true) = l.length := by simp
+
+@[simp] theorem countp_false {l : List α} : (l.countp fun _ => false) = 0 := by simp
+
+@[simp] theorem countp_map (p : β → Bool) (f : α → β) :
+    ∀ l, countp p (map f l) = countp (p ∘ f) l
+  | [] => rfl
+  | a :: l => by rw [map_cons, countp_cons, countp_cons, countp_map p f l]; rfl
+
+variable {p q}
+
+theorem countp_mono_left (h : ∀ x ∈ l, p x → q x) : countp p l ≤ countp q l := by
+  induction l with
+  | nil => apply Nat.le_refl
+  | cons a l ihl =>
+    rw [forall_mem_cons] at h
+    have ⟨ha, hl⟩ := h
+    simp [countp_cons]
+    cases h : p a
+    . simp
+      apply Nat.le_trans ?_ (Nat.le_add_right _ _)
+      apply ihl hl
+    . simp [ha h, Nat.add_one]
+      apply Nat.succ_le_succ
+      apply ihl hl
+
+theorem countp_congr (h : ∀ x ∈ l, p x ↔ q x) : countp p l = countp q l :=
+  Nat.le_antisymm
+    (countp_mono_left fun x hx => (h x hx).1)
+    (countp_mono_left fun x hx => (h x hx).2)
+
+end Countp
+
+/-! ### count -/
+
+section Count
+
+variable [DecidableEq α]
+
+@[simp] theorem count_nil (a : α) : count a [] = 0 := rfl
+
+theorem count_cons (a b : α) (l : List α) :
+    count a (b :: l) = count a l + if a = b then 1 else 0 := by conv =>
+  simp [count, countp_cons]
+  lhs
+  simp only [eq_comm]
+
+@[simp] theorem count_cons_self (a : α) (l : List α) : count a (a :: l) = count a l + 1 := by
+  simp [count_cons]
+
+@[simp] theorem count_cons_of_ne (h : a ≠ b) (l : List α) : count a (b :: l) = count a l := by
+  simp [count_cons, h]
+
+theorem count_tail : ∀ (l : List α) (a : α) (h : 0 < l.length),
+      l.tail.count a = l.count a - if a = get l ⟨0, h⟩ then 1 else 0
+  | head :: tail, a, h => by simp [count_cons]
+
+theorem count_le_length (a : α) (l : List α) : count a l ≤ l.length :=
+  countp_le_length _
+
+theorem Sublist.count_le (h : l₁ <+ l₂) (a : α) : count a l₁ ≤ count a l₂ :=
+  h.countp_le _
+
+theorem count_le_count_cons (a b : α) (l : List α) : count a l ≤ count a (b :: l) :=
+  (sublist_cons _ _).count_le _
+
+theorem count_singleton (a : α) : count a [a] = 1 := by simp
+
+theorem count_singleton' (a b : α) : count a [b] = if a = b then 1 else 0 := by
+  simp [count_cons]
+
+@[simp] theorem count_append (a : α) : ∀ l₁ l₂, count a (l₁ ++ l₂) = count a l₁ + count a l₂ :=
+  countp_append _
+
+theorem count_concat (a : α) (l : List α) : count a (concat l a) = succ (count a l) := by simp
+
+@[simp] theorem count_pos {a : α} {l : List α} : 0 < count a l ↔ a ∈ l := by
+  simp only [count, countp_pos, beq_iff_eq, exists_eq_right]
+
+@[simp] theorem one_le_count_iff_mem {a : α} {l : List α} : 1 ≤ count a l ↔ a ∈ l := count_pos
+
+theorem count_eq_zero_of_not_mem {a : α} {l : List α} (h : a ∉ l) : count a l = 0 :=
```

Comment:
This was marked `simp` in mathlib, but in Std causes the `simpNF` linter to complain because the LHS of `count_nil` is no longer simp-normal. Thoughts on the right way to proceed here?

Reply:
Resolved by setting `count_nil` priority higher.